WordPress/MediaEditor/MediaEditorHub.swift

```diff
@@ -78,16 +117,128 @@ class MediaEditorHub: UIViewController {
+        let imageAvailable = availableThumbs[selectedThumbIndex] ?? availableImages[selectedThumbIndex]
+
+        imageAvailable == nil ? showActivityIndicator() : hideActivityIndicator()
     }
 
     static func initialize() -> MediaEditorHub {
         return UIStoryboard(name: "MediaEditorHub", bundle: nil).instantiateViewController(withIdentifier: "hubViewController") as! MediaEditorHub
     }
 
+    private enum Constants {
+        static var thumbCellIdentifier = "thumbCell"
+        static var imageCellIdentifier = "imageCell"
+    }
+}
+
+// MARK: - UICollectionViewDataSource
+
+extension MediaEditorHub: UICollectionViewDataSource {
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return numberOfThumbs
+    }
+
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        // Thumbs Collection View
+        if collectionView == thumbsCollectionView {
+            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: Constants.thumbCellIdentifier, for: indexPath) as? MediaEditorThumbCell else {
+                return UICollectionViewCell()
+            }
+
+            cell.thumbImageView.image = availableThumbs[indexPath.row]
+            indexPath.row == selectedThumbIndex ? cell.showBorder(color: selectedColor) : cell.hideBorder()
+
+            return cell
+        }
+
+        // Images Collection View
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: Constants.imageCellIdentifier, for: indexPath) as? MediaEditorImageCell else {
+            return UICollectionViewCell()
+        }
+
+        cell.imageView.image = availableImages[indexPath.row] ?? availableThumbs[indexPath.row]
+
+        showOrHideActivityIndicator()
+
+        return cell
+    }
+}
+
+// MARK: - UICollectionViewDelegateFlowLayout
+
+extension MediaEditorHub: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: imagesCollectionView.frame.width, height: imagesCollectionView.frame.height)
+    }
+}
+
+// MARK: - UICollectionViewDelegate
+
+extension MediaEditorHub: UICollectionViewDelegate {
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        guard collectionView == thumbsCollectionView else {
+            return
+        }
+
+        selectedThumbIndex = indexPath.row
+        imagesCollectionView.scrollToItem(at: indexPath, at: .right, animated: true)
+    }
+
+    func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        guard scrollView == imagesCollectionView, isUserScrolling else {
+            return
+        }
+
+        let index = Int(round(scrollView.bounds.origin.x / imagesCollectionView.frame.width))
```

Comment:
I'm not sure if I'm reading this right, but if I am, you can use this instead https://developer.apple.com/documentation/uikit/uicollectionview/1618020-indexpathsforvisibleitems

```swift
thumbsCollectionView.indexPathsForVisibleItems[0]
```

Reply:
Here we're calculating which thumb should be selected based on the scroll in the `imagesCollectionView`.

You can quickly test this by slowing swiping to the next image. When you're seeing more than 50% of the next cell, the corresponding thumb is automatically selected.

But this code is a little bit confusing. I will refactor it to make it more meaningful.

Reply:
Oh, now that I'm reading again I guess your suggestion was to use:

```swift
thumbsCollectionView.indexPathsForVisibleItems[0]
```

However, I tried it and it doesn't work as I intended. To move to the correct thumbnail I'll need to check the direction of the scroll and check the `indexPathsForVisibleItems`.

Reply:
Changed in a536e331ebf8d94e1c327451a6e5655405c42a3b.

Let me know what you think.

Reply:
Looks fine to me.